identity/aziot-identityd/src/lib.rs

```diff
@@ -702,76 +634,199 @@ impl Api {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let public_key = key_engine.load_public_key(&key_handle).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+
+                    (private_key, private_key_string, public_key)
+                }
+                KeyType::Raw => {
+                    let rsa = openssl::rsa::Rsa::generate(2048).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let private_key = openssl::pkey::PKey::from_rsa(rsa).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let private_key_pem =
+                        private_key.private_key_to_pem_pkcs8().map_err(|err| {
+                            Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                        })?;
+                    let private_key_pem =
+                        std::string::String::from_utf8(private_key_pem).map_err(|err| {
+                            Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                        })?;
+                    let public_key = private_key.public_key_to_pem().map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let public_key = openssl::pkey::PKey::public_key_from_pem(&public_key)
+                        .map_err(|err| {
                             Error::Internal(InternalError::CreateCertificate(Box::new(err)))
                         })?;
 
-                    (identity_public_key, identity_private_key)
-                };
-
-                let csr = create_csr(
-                    &subject,
-                    &identity_public_key,
-                    &identity_private_key,
-                    attributes,
-                )
-                .map_err(|err| Error::Internal(InternalError::CreateCertificate(Box::new(err))))?;
-                csr
+                    (private_key, private_key_pem, public_key)
+                }
             };
 
-            let _new_cert = self
-                .cert_client
-                .create_cert(&identity_cert, &csr, None)
-                .await
-                .map_err(|err| Error::Internal(InternalError::CreateCertificate(Box::new(err))))?;
-        }
+            let result = async {
+                let csr =
+                    create_csr(&subject, &public_key, &private_key, attributes).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+
+                let cert = self
+                    .cert_client
+                    .create_cert(&identity_cert, &csr, None)
```

Comment:
What's your issuer for these local identities? I thought all the certs would chain to an independent local identities CA certificate.

Reply:
Yes, the issuer is on an EST server. The CSR is sent to CS, which then makes the EST request to sign.